### PR TITLE
log_backup: make a more rusty `CallbackWaitGroup`

### LINKED ISSUE #16739: log_backup: the type `CallbackWaitGroup` is error-prone
### What version of TiKV are you using?
Current master.

### What operating system and CPU are you using?
macOS at an aarch64 CPU, but it should be independent to CPU arch.

### Steps to reproduce
Run this:

```fish
for i in (seq 100)
   if ! cargo test --package backup-stream --lib -- utils::test::test_wait_group --exact
       break
   end
   echo "passed $i"
end
```

### What did you expect?
This test should always success -- no race conditions here.

### What did happened?
It will fail after about 20 times of trying.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

Issue Number: Close #16739

What's Changed:

```commit-message
This make `CallbackWaitGroup` returns an equivalent future of the `BoxFuture` returned by `wait`. Also this fixed where a stale notify may also resolve the future.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
I ran the task for 100 times. No error happens anymore.
```fish
for i in (seq 100)
   if ! cargo test --package backup-stream --lib -- utils::test::test_wait_group --exact
       break
   end
   echo "passed $i"
end
```

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Fixed a bug where make there are a slight possibility that log backup lost a few records.
```

### CODE REVIEW

File: components/backup-stream/src/utils.rs [hunk 1]
```diff
@@ -379,47 +379,61 @@ pub fn should_track_lock(l: &Lock) -> bool {
...
+
+pub struct WaitAll<'a>(&'a FutureWaitGroup);
+
+impl<'a> Future for WaitAll<'a> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Fast path: nothing to wait.
+        let running = self.0.running.load(Ordering::SeqCst);
+        if running == 0 {
```
Comment (3pointer) [inline components/backup-stream/src/utils.rs:403]: [("add_work", 0), ("work_done", 1), ("wait", 0), ("add_work", 1)]

will this wait return early before the second add work?

Comment (YuJuncen) [reply components/backup-stream/src/utils.rs:403]: Yes, and it is the caller's responsibility to call wait after all works has been spawn.

**Review by Leavrth — APPROVED** [review]

**Review by iosmanthus — APPROVED** [review]

**Pull request merged**